src/CrossChainReceiverFactory.sol

```diff
@@ -0,0 +1,406 @@
+                hex"60265f8160095f39f35f5f365f5f37365f6c",
+                bytes13(uint104(uint160(address(this)))),
+                hex"5af43d5f5f3e6022573d5ffd5b3d5ff3"
+            )
+        );
+    }
+
+    modifier onlyProxy() {
+        require(this != _cachedThis);
+        _;
+    }
+
+    modifier noDelegateCall() {
+        require(this == _cachedThis);
+        _;
+    }
+
+    modifier onlyFactory() {
+        require(msg.sender == address(_cachedThis));
+        _;
+    }
+
+    function _verifyRoot(bytes32 root, address originalOwner) internal view returns (bool result) {
+        bytes32 initHash = _proxyInitHash;
+        CrossChainReceiverFactory factory = _cachedThis;
+        assembly ("memory-safe") {  
+            let ptr := mload(0x40)
+
+            // derive creation salt
+            mstore(0x14, originalOwner)
+            mstore(0x00, root)
+            let salt := keccak256(0x00, 0x34)
+
+            // 0xff + factory + salt + hash(initCode)
+            mstore(0x4d, initHash)
+            mstore(0x2d, salt)
+            mstore(0x0d, factory)
+            mstore(0x00, 0xff00000000000000)
+            let computedAddress := keccak256(0x18, 0x55)
+
+            // restore clobbered memory
+            mstore(0x60, 0x00)
+            mstore(0x40, ptr)
+
+            // verify that `salt` was used to deploy `address(this)`
+            result := iszero(shl(0x60, xor(address(), computedAddress)))
+        }
+    }
+
+    // @inheritdoc IERC1271
+    function isValidSignature(bytes32 hash, bytes calldata signature)
+        external
+        view
+        override
+        /* `_verifyRoot` hashes `_cachedThis`, making this function implicitly `onlyProxy` */
+        returns (bytes4)
+    {
+        { // Merkle proof validation
+            address owner;
+            bool validOwner;
+            bytes32 leaf;
+            bytes32[] calldata proof;
+
+            assembly ("memory-safe") {
+                // This assembly block is equivalent to:
+                //     leaf = keccak256(abi.encode(hash, block.chainid));
+                // except that it's cheaper and doesn't allocate memory. We make the assumption here that
+                // `block.chainid` cannot alias a valid tree node or signing hash. Realistically,
+                // `block.chainid` cannot exceed 2**53 - 1 or it would cause significant issues elsewhere in
+                // the ecosystem. This also means that the sort order of the hash and the chainid is
+                // backwards from what `MerkleProofLib` produces, again protecting us against extension
+                // attacks.
+                mstore(0x00, hash)
+                mstore(0x20, chainid())
+                leaf := keccak256(0x00, 0x40)
+
+                // Following assembly blocks are equivalent to:
+                //     (owner, proof) = abi.decode(signature, (address, bytes32[]));
+                // except we omit all the range and overflow checking.
```

Comment:
This assembly block doesn't match the code below. Also this needs an explanation around the assumptions that make `validOwner` a good semantic fit. In other words, under what circumstances might the encoding for a merkle signature collide with a 7739 signature?

Reply:
Tried to use "Following assembly block**s**" (with the s) and another comment in the next block saying "continuation" 😅